doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,214 @@
+
+The `DROP SECRET` statement drops an existing secret. It returns an error
+if a secret with the specified name does not exist unless the `IF EXIST`
+clause is present. Dropping a secret that is in use by a `CONNECTOR` is not
+permitted unless `CASCADE` is present.
+
+If the [secrets controller](#secrets-controller) in use supports multi-operation
+atomicity, a SQL transaction can contain any number of secret DDL statements;
+otherwise a SQL transaction can contain only one secret DDL statement. In either
+case, the SQL transaction must not contain any other type of statements. The
+secret DDL statements are applied atomically to the secrets controller if and
+only if the transaction commits.
+
+Note that while the catalog updates to a secret apply immediately, changes to
+the contents of the secret do not propagate to downstream consumers of the
+secret on any particular timeline.
+
+The `SHOW SECRETS` statement lists the *names* of the secrets in the specified
+schema, or the first schema in the search path if no schema is explicitly
+specified, optionally filtered by the provided `LIKE` pattern or `WHERE`
+expression. The filter clauses which work analogously to the same clauses in the
+`SHOW DATABASES` statement.
+
+Note that there is intentionally no way to view the contents of secrets.
+
+### System catalog changes
+
+A new `mz_secrets` table with the following structure will describe the
+available secrets in the system:
+
+Field       | Type      | Meaning
+------------|-----------|--------
+`id`        | `bigint`  | The ID of the secret.
+`schema_id` | `bigint`  | The ID of the schema to which the secret belongs.
+`name`      | `text`    | The name of the secret.
+
+Note that, again, the contents of the secrets are not exposed.
+
+### Referencing secrets
+
+There is presently no way to use a secret. The existing `CREATE SOURCE` and
+`CREATE SINK` commands will be extended to support referencing secrets, but
+that work is left to a future design document.
+
+## Implementation
+
+### Catalog changes
+
+The [`CatalogItem`] enum will be extended with an additional `Secret` variant:
+
+```rust
+pub enum CatalogItem {
+    // ...
+    Secret,
+}
+```
+
+Unlike the other catalog items, there is no data associated with a secret. The
+name and ID of the secret are stored in the `CatalogEntry` struct, and the
+actual data for the secret will be stored in a secret backend (see next
+section). Therefore there is presently no additional metadata that needs to be
+recorded about a secret.
+
+When serializing a secret, its `create_sql` will be synthesized as:
+
+```sql
+CREATE SECRET <name> AS '******'
+```
+
+When deserializing a secret on bootstrap, the `'******'` will be ignored.
+
+### Secrets controller
+
+The `Coordinator` will host a `SecretsController` that securely handles the
+storage of secrets. It will interact with the controller via the following
+interface.
+
+```rust
+/// Securely stores secrets.
+pub trait SecretsController {
```

Comment:
The system as described can end up with secrets in the controller that are missing from the catalog and you can only clean them up with administrative actions.

The easiest way I see that doesn't involve the coordinator durably writing down secrets before updating the secret controller (which is scary) is to add a `list` method here so that on bootstrap we can reconcile any secrets that got lost.

Reply:
Good point! I added a whole section on the complex integration between catalog transactions and secrets controller transactions, which invovles a `list()` method that can be used to periodically garbage collect orphaned secrets.